feat: Readonly and inline ibm-label component for Input and TextArea

anemonetea opened the pull request:
Closes:
carbon-design-system/carbon-components-angular#2291

Changing the `ibm-label` component to add inputs for `inline` and `readonly` options to match what is being done in React:
- React: https://github.com/carbon-design-system/carbon/blob/v10/packages/react/src/components/TextInput/TextInput.js
- Scss: https://github.com/carbon-design-system/carbon/blob/v10/packages/styles/scss/components/text-input/_text-input.scss

#### Changelog

**New**

* Inline styling of `ibm-label` controlled by `[inline]` input
* Readonly styling of `ibm-label` controlled by `[readonly]` input
* Readonly behaviour for `input` and `textarea` elements added to `ibmText` and `ibmTextArea` directives using `[readonly]` input
* StoryBook example of input components that use templates as `ibm-label` helper text inputs

**Changed**

* Fixed unit tests for input component

**Removed**
N/A

Akshat55 commented on src/input/label.component.ts:
```diff
@@ -148,6 +174,7 @@ export class Label implements AfterContentInit, AfterViewInit {
 	@ContentChild(TextArea, { static: false }) textArea: TextArea;
 
 	@HostBinding("class.bx--form-item") labelClass = true;
+	@HostBinding("class.bx--text-input-wrapper") inputWrapperClass = true;
```
Is this intentional? It will be assigned to the host in both text area & text input or no matter what the content is 🤔

anemonetea replied:
It's what makes the `inline` variant work... I currently have it as a viable option for textarea as well.

anemonetea replied:
`${prefix}--text-input--${size}` is set in the `ibmText` directive:
https://github.com/carbon-design-system/carbon-components-angular/blob/master/src/input/input.directive.ts#L26-L34

Akshat55 commented on src/input/label.component.ts:
```diff
@@ -16,77 +16,90 @@ import { TextArea } from "./text-area.directive";
+						'bx--text-area__invalid-icon': textArea
+					}">
+				</svg>
+				<svg
+					*ngIf="!invalid && warn && !readonly"
+					ibmIcon="warning--alt--filled"
+					size="16"
+					class="bx--text-input__invalid-icon bx--text-input__invalid-icon--warning">
+				</svg>
+				<svg
+					*ngIf="readonly"
+					ibmIcon="edit--off"
+					size="16"
+					class="bx--text-input__readonly-icon">
+				</svg>
+				<ng-content select="input,textarea,div"></ng-content>
+			</div>
+			<ng-container *ngIf="!inline"
+				[ngTemplateOutlet]="helperTextTemplate" [ngTemplateOutletContext]="{inline: inline}">
+			</ng-container>
 		</div>
+
+		<ng-template #helperTextTemplate let-inline=inline>
+			<div *ngIf="!skeleton && helperText && ((!invalid && !warn) || inline)"
+				class="bx--form__helper-text"
+				[ngClass]="{
+					'bx--form__helper-text--disabled': disabled
+				}">
+				<ng-container *ngIf="!isTemplate(helperText)">{{helperText}}</ng-container>
+				<ng-template *ngIf="isTemplate(helperText)" [ngTemplateOutlet]="helperText"></ng-template>
+			</div>
+			<div *ngIf="!inline && !warn && invalid" class="bx--form-requirement">
+				<ng-container *ngIf="!isTemplate(invalidText)">{{invalidText}}</ng-container>
+				<ng-template *ngIf="isTemplate(invalidText)" [ngTemplateOutlet]="invalidText"></ng-template>
+			</div>
+			<div *ngIf="!inline && !invalid && warn" class="bx--form-requirement">
+				<ng-container *ngIf="!isTemplate(warnText)">{{warnText}}</ng-container>
+				<ng-template *ngIf="isTemplate(warnText)" [ngTemplateOutlet]="warnText"></ng-template>
+			</div>
+		<ng-template>
```
I don't think this template takes `TextArea` into account. It's applying classes designed for TextInput to the entire label component. if the styles are changed on carbon's end, this would be a breaking change. Perhaps it would be best if you create a template for inline & swap to default if not? Thoughts?

Akshat55 replied:
Just an FYI, Textarea does NOT support inline styling, It _WORKS_, but we shouldn't add a feature like this if it is not in Carbon guideline.

anemonetea replied:
Oh Ok, I actually wasn't sure whether textarea was supposed to have the inline styling or not... I asked it in the issue as well -- https://github.com/carbon-design-system/carbon-components-angular/issues/2291#issuecomment-1458418523
So, I'll remove the textarea part.

Akshat55 replied:
Oh sorry! I didn't see the notification for that issue.

Akshat55 reviewed (changes requested):
Feature could use some polishing, we want to make sure we don't break any existing components.

anemonetea commented on src/input/input.directive.ts:
```diff
@@ -38,4 +38,20 @@ export class TextInput {
+		// Set attribute based on input
+		if (this._readonly === true) {
+			this.readonlyAttr = "true";
+		} else if (this._readonly === false) {
+			this.readonlyAttr = null;
+		}
```
I'm not sure about this... because the native `readonly` attribute would act as if it was `true` if it was set to some random string, for example. If we make our `readonly` act as if it's `false` when getting a random string as input, then that would be even more confusing, hahah...

If I were using this on the other end, I think I'd want to see an error being thrown if I'm passing in the wrong type of thing. Typically, typescript should handle this, but I understand if not everybody has it fully working, especially for Carbon.

anemonetea replied:
Actually, never mind... because the `ibm-label` input for `readonly` works like `==` rather than `===` to apply the style:
`@HostBinding("class.bx--text-input-wrapper--readonly") readonly`

anemonetea commented:
Addressed comments:
 - remove `readonly` variant from TextArea version
 - use `if (_readonly) { set attr to "true" } else { set attr to null }` instead of relying on `(_readonly === true)`
 - import a Carbon icon into the story instead of copy-pasting them
 - other minor comments

Akshat55 reviewed (changes requested):
I have tested it with my requested changes, but let me know if anything isn't working in the template OR my changes in the class & I'll address it.

anemonetea commented on src/input/label.component.ts:
```diff
@@ -106,12 +119,25 @@ export class Label implements AfterContentInit, AfterViewInit {
 	 * Set to `true` for a disabled label.
 	 */
 	@Input() disabled = false;
+	/**
+	 * Set to `true` for a read-only label.
+	 */
+	@Input() @HostBinding("class.bx--text-input-wrapper--readonly") readonly = false;
+	/**
+	 * Set to `true` for an inline style label.
+	 */
+	@Input() @HostBinding("class.bx--text-input-wrapper--inline") inline = false;
```
I just want to point out that all of this extra logic still doesn't account for the case of a `div` being in the content of `ibm-label`:
`<ng-content select="input,textarea,div">`.

This all sounds like a bigger problem, if you ask me... Just the fact that carbon-ng uses the common label component whereas React doesn't (and we share the styles) -- it's not scalable. So, it is probably going to break down the line... and I'm afraid to even test the `div` version 😅 (there are no preexisting tests or storybooks for the div version...)

anemonetea replied:
Some notes about the current `.bx--text-input-wrapper`:
- never does anything on its own, only in combination with `-inline` classes
- relies on `.bx--form-item` to set the `display: flex` for the inline styling of `flex-wrap` to work
![Screenshot 2023-04-21 at 4 49 43 PM](https://user-images.githubusercontent.com/26636906/233732461-c108bdf7-327b-4772-adea-3910b0c38442.png)

anemonetea replied:
`tslint` doesn't like the getter-setter code :( 
```
ERROR: src/input/label.component.ts:159:2 - Declaration of public instance field not allowed after declaration of protected instance field. Instead, this should come after public static fields.
```

anemonetea replied:
I've changed this...

anemonetea replied:
I created a CodeSandbox with a bunch of different types of contents in `ibm-label`, including ones meeting the `div` selector: https://codesandbox.io/s/carbon-ng-inputs-1sf4f3?file=/src/app/app.component.html
 - note that the `div` version ends up with `bx--text-input__field-wrapper`, and so does textArea for some reason...

anemonetea commented on src/icon/icon.module.ts:
```diff
@@ -55,6 +55,7 @@ import Warning16 from "@carbon/icons/es/warning/16";
 import WarningFilled16 from "@carbon/icons/es/warning--filled/16";
 import WarningFilled20 from "@carbon/icons/es/warning--filled/20";
 import WarningAltFilled16 from "@carbon/icons/es/warning--alt--filled/16";
+import EditOff16 from "@carbon/icons/es/edit--off/16";
```
done

anemonetea commented on src/icon/icon.module.ts:
```diff
@@ -132,7 +133,8 @@ export class IconModule {
 			WarningFilled20,
-			WarningAltFilled16
+			WarningAltFilled16,
+			EditOff16
```
done

anemonetea commented on src/input/input.directive.ts:
```diff
@@ -38,4 +38,20 @@ export class TextInput {
+		return this._readonly;
+	}
+	private _readonly: boolean;
+
```
I could swear I thought I tested this at some point... but now, you're right -- it's not actually needed. 🤯 
So, I've removed it...

anemonetea commented on src/input/label.component.ts:
```diff
@@ -7,89 +7,104 @@ import {
+				class="bx--form__helper-text"
+				[ngClass]="{
+					'bx--form__helper-text--disabled': disabled,
+					'bx--form__helper-text--inline': inline
```
I had missed adding this `--inline` class to helper text... 
- https://github.com/carbon-design-system/carbon/blob/v10/packages/react/src/components/TextInput/TextInput.js#L111-L114

anemonetea commented on src/input/label.component.ts:
```diff
@@ -16,77 +16,90 @@ import { TextArea } from "./text-area.directive";
+				<ng-container *ngIf="!isTemplate(warnText)">{{warnText}}</ng-container>
+				<ng-template *ngIf="isTemplate(warnText)" [ngTemplateOutlet]="warnText"></ng-template>
+			</div>
+		<ng-template>
```
I've created another branch to test this out and put it in a PR (in my fork) for viewing the diffs: https://github.com/anemonetea/carbon-components-angular/pull/1/files

(From another comment) I agree that the outer wrapper shouldn't have been getting applied to in the `textArea` variant, so my branch fixes that issue.

Also agree that the shared `ibm-label` wrapper is making this code messy and it's not the best for maintenance...

> We want to make this change in a way where we don't require existing CCA users to update their styles or make any changes. Changing order of label, adding wrapper classes will cause a breaking change for users that are making targeted style changes to label or input for example.

My concern with this approach is that we're further deviating from how React uses these CSS classes by not using `bx--text-input__field-outer-wrapper` for regular (not inline) instances of the `textInput` variant. So, if they change how the CSS works, our Angular stuff might break...

> we don't require existing CCA users to update their styles or make any changes.

At least in my experience, my product team follows the wisdom of avoiding modifications of internal Carbon classes just in case Carbon makes changes. So, it's Carbon's prerogative to make changes to the internal working of Carbon components. They're also scared of updates because they might break stuff... so, every update needs to trigger regression testing. At the end of the day, it sounds like more of a matter of setting expectations about what kind of versioning change (fix, minor, major) can be expected to make changes to the internal DOM / CSS structure of components, and which ones can't -- not sure about how you guys handle this.

Akshat55 commented:
Hey @anemonetea  - we recently added new input/textarea labels to make it easier to add input/textarea specific styling. Could you close this and remake the PR?

It will be a lot easier to apply changes now. You will need to update `text-input-label.component.ts`. Just an FYI, we just released CCA v5, so you will need to make the PR into v10 branch. 

Let me know if you have any questions and I can guide you through the changes. Feel free to reach out to me via slack!

Akshat55 commented:
Closing this - it makes sense to create multiple new PRs per feature. Should be much easier since we have separate components for text input and text area now.
